lib/remap-lang/src/type_def.rs

```diff
@@ -216,20 +196,149 @@ impl TypeDef {
     }
 }
 
+/// The inner type defnition for a type.
+///
+/// Maps will have an inner type definition that represents the typedefs of the
+/// fields contained by the map.
+///
+/// Arrays have a single inner type definition representing the type for each element
+/// contained within the array.
```

Comment:
There isn't really a specific technical reason, but I'm not convinced it would be particularly useful. I imagine every array in the system will be in cases where the indexes won't be known at compile time.

I can't think of a scenario where someone is going to be setting a specific index of an array with a known map. It's always going to be the results of function calls where the number of rows returned are not known until run time.

So rather than spend effort working on this case, I think we should figure out a better way to handle it. So we know, for example, than `parse_regex_all` will return an Array of Maps - and we know what fields those Maps will contain. So if we can provide a way of guaranteeing that the script is accessing a row in that Array that has been populated, we will know the `type_def` of the fields in that row.

For example if we had a: 

```coffeescript
for each row in parse_regex_all(.thing, /.../) {
  # We know the precise type_def of row here.
}
```

I think that is probably the scenario that is most likely to be used and we should look to getting this to work.

Reply:
I agree, array literals are way less useful currently.

Let's at least file an issue to track this, since it's a hole in our type checking that we can fill if we have the time or enough feedback that it's useful to do so.

Once (_if_) we have support for iterating collections (https://github.com/timberio/vector/pull/5875), I can see someone wanting to do:

```rust
grok_patterns = [
	"foo",
	"bar",
]

for pattern in grok_patterns {
	parse_grok(pattern)
}
```

at which point it'd be nice to not fail due to an unknown function argument kind.

Reply:
Have raised #6297 to track this.